osquery/tables/system/linux/elf_info.cpp

```diff
@@ -0,0 +1,219 @@
+
+      Row r;
+      r["path"] = path;
+      auto gnu_type =
+          kGNUTypes.find(static_cast<elf::ElfTypes::Word>(hdr.type));
+      if (gnu_type != kGNUTypes.end()) {
+        r["name"] = gnu_type->second;
+      } else {
+        r["name"] = to_string(hdr.type);
+      }
+      r["offset"] = std::to_string(hdr.offset);
+      r["vaddr"] = std::to_string(hdr.vaddr);
+      r["flags"] = to_string(hdr.flags);
+      r["psize"] = std::to_string(hdr.filesz);
+      r["msize"] = std::to_string(hdr.memsz);
+      r["align"] = std::to_string(hdr.align);
+      results.push_back(r);
+    }
+  };
+
+  genElfInfo(context, lambda);
+  return results;
+}
+
+QueryData getELFSymbols(QueryContext& context) {
+  QueryData results;
+
+  auto lambda = [&results](const elf::elf& f, const std::string& path) {
+    for (auto& sec : f.sections()) {
+      auto& hdr = sec.get_hdr();
+
+      if (hdr.type != elf::sht::symtab && hdr.type != elf::sht::dynsym) {
+        continue;
+      }
+
+      Row r;
+      r["path"] = path;
+      r["table"] = sec.get_name();
+
+      for (const auto& sym : sec.as_symtab()) {
+        auto& d = sym.get_data();
+        r["addr"] = std::to_string(d.value);
+        r["size"] = std::to_string(d.size);
+        r["type"] = to_string(d.type());
+        r["binding"] = to_string(d.binding());
+        r["offset"] = to_string(d.shnxd);
+        r["name"] = sym.get_name();
+        results.push_back(r);
+      }
+    }
+  };
+
+  genElfInfo(context, lambda);
+  return results;
+}
+
+QueryData getELFSections(QueryContext& context) {
+  QueryData results;
+
+  auto lambda = [&results](const elf::elf& f, const std::string& path) {
+    for (auto& sec : f.sections()) {
+      auto& hdr = sec.get_hdr();
+
+      Row r;
+      r["path"] = path;
+
+      // -d == dynamic section!
+      printf(
+          "section: %s type=%d\n", sec.get_name().c_str(), sec.get_hdr().type);
+
+      r["name"] = sec.get_name();
+      r["type"] = std::to_string(static_cast<elf::ElfTypes::Word>(hdr.type));
+      r["addr"] = std::to_string(hdr.addr);
+      r["offset"] = std::to_string(hdr.offset);
+      r["size"] = std::to_string(hdr.size);
+      r["flags"] = to_string(hdr.flags);
+      r["link"] = to_string(hdr.link);
+      r["align"] = std::to_string(hdr.addralign);
+      results.push_back(r);
+    }
+  };
+
+  genElfInfo(context, lambda);
+  return results;
+}
+
+QueryData getELFDynamic(QueryContext& context) {
+  QueryData results;
+
+  auto lambda = [&results](const elf::elf& f, const std::string& path) {
+    for (auto& sec : f.sections()) {
+      auto& hdr = sec.get_hdr();
+      if (hdr.type != elf::sht::dynamic) {
+        continue;
+      }
+
+      Row r;
+      r["path"] = path;
+      const auto* data = sec.data();
+      if (f.get_hdr().ei_class == elf::elfclass::_32) {
+        r["class"] = "32";
+        auto* dynamic = reinterpret_cast<const Elf32_Dyn*>(data);
+        for (const auto* d = dynamic; d->d_tag != DT_NULL; ++d) {
```

Comment:
Hm, this seems to be the same `for` as below. Not sure how, but perhaps this could be restructured so both `32` and `64` use the same `for` loop? 
```
r["class"] = f.get_hdr().ei_class == elf::elfclass::_32 ? "32" : "64";
auto* dynamic = r["class"] == "32" ? reinterpret_cast<const Elf32_Dyn*>(data) : reinterpret_cast<const Elf64_Dyn*>(data);
for (...) {
  ...
```
That however doesn't seem nearly as readable to me.

Reply:
Agree, but this doesn't quite work. The ternary statement must return the same type.